ports/zephyr/machine_uart.c

```diff
@@ -60,16 +73,30 @@ static void mp_machine_uart_print(const mp_print_t *print, mp_obj_t self_in, mp_
 }
 
 static void mp_machine_uart_init_helper(machine_uart_obj_t *self, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
-    enum { ARG_timeout, ARG_timeout_char };
+    enum { ARG_txbuf, ARG_rxbuf, ARG_timeout, ARG_timeout_char };
     static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_txbuf, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = UART_RX_RING_BUF_DEF_SIZE} },
+        { MP_QSTR_rxbuf, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = UART_TX_RING_BUF_DEF_SIZE} },
         { MP_QSTR_timeout, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0} },
         { MP_QSTR_timeout_char, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0} },
     };
+
     mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
     mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
 
+    self->txbuf_size = args[ARG_txbuf].u_int;
+    self->rxbuf_size = args[ARG_rxbuf].u_int;
     self->timeout = args[ARG_timeout].u_int;
     self->timeout_char = args[ARG_timeout_char].u_int;
+
+    self->tx_buffer = m_tracked_calloc(self->txbuf_size, sizeof(uint8_t));
+    ring_buf_init(&self->tx_ringbuffer, self->txbuf_size, self->tx_buffer);
```

Comment:
Any reason not to use MicroPython's ringbuffer (see `py/ringbuf.h`)?  I guess it doesn't really matter much, I'm just more familiar with our version rather than zephyr's.

With zephyr's implementation, if you allocate N bytes, does it only give you N-1 slots in the ringbuffer?

Reply:
I forgot we had our own implementation of a ringbuffer. I'll switch to it.